src/data/polynomial/partial_fractions.lean

```diff
@@ -0,0 +1,132 @@
+  ∃ q r₁ r₂ : R[X], r₁.degree < g₁.degree ∧ r₂.degree < g₂.degree ∧
+  (↑f : K) / (↑g₁ * ↑g₂) = ↑q + ↑r₁ / ↑g₁ + ↑r₂ / ↑g₂ :=
+begin
+  rcases hcoprime with ⟨ c, d, hcd ⟩,
+  refine ⟨ (f*d) /ₘ g₁ + (f*c) /ₘ g₂ , (f*d) %ₘ g₁ , (f*c) %ₘ g₂ ,
+    (degree_mod_by_monic_lt _ hg₁) , (degree_mod_by_monic_lt _ hg₂) , _⟩,
+  have hg₁' : (↑g₁ : K) ≠ 0,
+  { norm_cast, exact hg₁.ne_zero_of_ne zero_ne_one, },
+  have hg₂' : (↑g₂ : K) ≠ 0,
+  { norm_cast, exact hg₂.ne_zero_of_ne zero_ne_one, },
+  have hfc := mod_by_monic_add_div (f * c) hg₂,
+  have hfd := mod_by_monic_add_div (f * d) hg₁,
+  field_simp,
+  norm_cast,
+  linear_combination (-1) * f * hcd + (-1) * g₁ * hfc + (-1) * g₂ * hfd,
+end
+
+end two_denominators
+
+section n_denominators
+
+open_locale big_operators classical
+
+/--
+Let R be an integral domain and f ∈ R[X]. Let s be a finite index set.
+Then, a fraction of the form f / ∏ (g i) can be rewritten as q + ∑ (r i) / (g i), where
+deg(r i) < deg(g i), provided that the g i are monic and pairwise coprime.
+-/
+lemma div_eq_quo_add_sum_rem_div (f : R[X]) {ι : Type*} {g : ι → R[X]} {s : finset ι}
+  (hg : ∀ i ∈ s, (g i).monic)
+  (hcop : set.pairwise ↑s (λ i j, is_coprime (g i) (g j))) :
+  ∃ (q : R[X]) (r : ι → R[X]), (∀ i ∈ s, (r i).degree < (g i).degree) ∧
+  (↑f : K) / ∏ i in s, ↑(g i) = ↑q + ∑ i in s, ↑(r i) / ↑(g i) :=
+begin
+  induction s using finset.induction_on with a b hab Hind f generalizing f,
+  { refine ⟨f, (λ (i : ι), (0 : R[X])), λ i, _, by simp⟩,
+    rintro ⟨⟩, },
+  { obtain ⟨q₀, r₁, r₂, hdeg₁, hdeg₂, (hf : (↑f : K) / _ = _)⟩ :=
+      div_eq_quo_add_rem_div_add_rem_div R K f
+      (_ : monic (g a))
+      (_ : monic ∏ (i : ι) in b, (g i))
+      _,
+    { obtain ⟨q, r, hrdeg, IH⟩ := Hind (λ i hi, hg i (finset.mem_insert_of_mem hi))
+        (set.pairwise.mono (show (b : set ι) ⊆ (insert a b : finset ι),
+          from finset.coe_subset.2 $ λ i hi, finset.mem_insert_of_mem hi) hcop) r₂,
+      refine ⟨q₀ + q, λ i, if i = a then r₁ else r i, _, _⟩,
+      { intro i,
+        split_ifs with h1,
+        { cases h1,
+          intro _,
+          exact hdeg₁, },
+        { intro hi,
+          refine hrdeg i (finset.mem_of_mem_insert_of_ne hi h1), }, },
+      norm_cast at ⊢ hf IH,
+      rw [finset.prod_insert hab, hf, IH, finset.sum_insert hab, if_pos rfl],
+      transitivity (↑(q₀ + q : R[X]) : K) + (↑r₁ / ↑(g a) + ∑ (i : ι) in b, ↑(r i) / ↑(g i)),
+      { push_cast, ring, },
+      congr' 2,
+      refine finset.sum_congr rfl (λ x hxb, _),
+      have hxa : ¬(x = a),
+      { rintro rfl,
+        exact hab hxb, },
+      rw if_neg hxa, },
+    { exact hg a (b.mem_insert_self a), },
+    { refine monic_prod_of_monic _ _ (λ i hi, hg i (finset.mem_insert_of_mem hi)), },
+    { refine is_coprime.prod_right (λ i hi, hcop (finset.mem_coe.2 (b.mem_insert_self a))
+       (finset.mem_coe.2 (finset.mem_insert_of_mem hi)) _),
+      rintro rfl,
+      exact hab hi, }, },
+end
```

Comment:
This proof uses some surprising constructions. Still, many of those are in my opinion better than the mathlib-typical style so I'm happy to learn that this is possible (like your hole-with-type-ascription in `obtain`).

A few minor stylistic tweaks:
 * if a tactic creates two goals and one of them is the "main" part of the proof, we don't need to put that between braces
 * `rw` will create new goals for any argument it can't figure out from the context, so we don't need the `have hxa`.
 * if `refine` solves a goal, we write `exact` 
```suggestion
begin
  induction s using finset.induction_on with a b hab Hind f generalizing f,
  { refine ⟨f, (λ (i : ι), (0 : R[X])), λ i, _, by simp⟩,
    rintro ⟨⟩, },
  obtain ⟨q₀, r₁, r₂, hdeg₁, hdeg₂, (hf : (↑f : K) / _ = _)⟩ :=
    div_eq_quo_add_rem_div_add_rem_div R K f
    (_ : monic (g a))
    (_ : monic ∏ (i : ι) in b, (g i))
    _,
  { obtain ⟨q, r, hrdeg, IH⟩ := Hind (λ i hi, hg i (finset.mem_insert_of_mem hi))
      (set.pairwise.mono ( finset.coe_subset.2 $ λ i hi, finset.mem_insert_of_mem hi) hcop) r₂,
    refine ⟨q₀ + q, λ i, if i = a then r₁ else r i, _, _⟩,
    { intro i,
      split_ifs with h1,
      { cases h1,
        intro _,
        exact hdeg₁, },
      { intro hi,
        exact hrdeg i (finset.mem_of_mem_insert_of_ne hi h1), }, },
    norm_cast at ⊢ hf IH,
    rw [finset.prod_insert hab, hf, IH, finset.sum_insert hab, if_pos rfl],
    transitivity (↑(q₀ + q : R[X]) : K) + (↑r₁ / ↑(g a) + ∑ (i : ι) in b, ↑(r i) / ↑(g i)),
    { push_cast, ring, },
    congr' 2,
    refine finset.sum_congr rfl (λ x hxb, _),
    rw if_neg,
    rintro rfl,
    exact hab hxb },
  { exact hg a (b.mem_insert_self a), },
  { exact monic_prod_of_monic _ _ (λ i hi, hg i (finset.mem_insert_of_mem hi)), },
  { refine is_coprime.prod_right (λ i hi, hcop (finset.mem_coe.2 (b.mem_insert_self a))
     (finset.mem_coe.2 (finset.mem_insert_of_mem hi)) _),
    rintro rfl,
    exact hab hi, },
end
```